diff -Naur ./sys/conf/NOTES /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/NOTES
--- ./sys/conf/NOTES	2008-02-23 21:08:16.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/NOTES	2008-02-23 19:59:46.000000000 +1100
@@ -2344,6 +2344,68 @@
 device		lpbb
 device		pcfclock
 
+# KGI - Kernel Graphic Interface
+#
+#
+# Enables KGI compatibility
+#
+options		KGI_COMPAT
+
+# Prevent KGI debug output
+#
+options		KGI_DBG_LEVEL=0
+
+# Enable probe of higher modes
+#
+options		KGC_RENDER_800x600
+options		KGC_RENDER_1024x768
+
+# Use 16bit mode
+#
+options		KGC_RENDER_16BITS
+
+# Enable console background
+#
+options		KGC_RENDER_BACKGROUND
+
+# The default font to be used by kgc console. See dev/kgc/kgc_gfbdflt.c
+# Currently nothing else than kgc_bold8x16 is in the kernel.
+#
+options		KGC_DEFAULT_FONT=kgc_bold8x16
+
+# KGI Kernel Input Parser
+#
+# It works on top of kbd in RAW mode. Scancodes are translated by
+# kip maps which are similar to Linux ones.
+#
+device		kip		# Default Kernel Input parser
+
+# Kernel Graphic Interface core
+#
+device		kgi		# Kernel Graphic Interface core system (include KII)
+
+# Input and Graphic user interfaces (e.g /dev/graphic and /dev/event)
+#
+device          kgu             # KGI graphic user interface
+device          kiu             # KGI input user interface
+
+# Default font to be used by kgc console. See dev/kgc/kgc_gfbdflt.c
+# Currently nothing else than kgc_bold8x16 is in the kernel.
+#
+options		KGC_DEFAULT_FONT=kgc_bold8x16
+
+# Kernel Graphic Console
+#
+device		kgc		# KGI console
+
+# Experimental FreeBSD adapter implementation on top of KGI
+#
+#device		kga		# FreeBSD compatible adapter (broken)
+
+# Syscons emulation for KGC
+#
+device		scemul
+
 # Kernel BOOTP support
 
 options 	BOOTP		# Use BOOTP to obtain IP address/hostname
diff -Naur ./sys/conf/files /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/files
--- ./sys/conf/files	2008-02-23 21:08:16.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/files	2008-02-23 19:59:46.000000000 +1100
@@ -758,6 +758,46 @@
 dev/joy/joy_isa.c		optional joy isa
 dev/joy/joy_pccard.c		optional joy pccard
 dev/kbdmux/kbdmux.c		optional kbdmux
+dev/kgc/kgc_emudumb.c		optional kgc kgi
+dev/kgc/kgc_emuxterm.c		optional kgc kgi
+dev/kgc/kgc_textbuf.c		optional kgc kgi
+dev/kgc/kgc_gfbdflt.c		optional kgc kgi
+dev/kgc/kgc_gfbrndr.c		optional kgc kgi
+dev/kgc/kgc_kgirndr.c		optional kgc kgi
+dev/kgc/kgc_mapper.c		optional kgc kgi
+dev/kgc/kgc_render.c		optional kgc kgi
+dev/kgc/kgc_backgnd.c		optional kgc kgi
+dev/kgc/kgc_bgndpcx.c		optional kgc kgi
+dev/kgc/kgc_daemon.c		optional kgc kgi
+dev/kgc/kgc_scroller.c		optional kgc kgi
+dev/kgc/kgc_textscroller.c	optional kgc kgi
+dev/kgc/render_if.m		optional kgc kgi
+dev/kgc/scroller_if.m		optional kgc kgi
+dev/scemul/sce_bootconsole.c	optional scemul kgc kgi
+dev/scemul/sce_vtconsole.c	optional scemul kgc kgi
+dev/scemul/sce_sysmouse.c	optional scemul kgc kgi
+dev/scemul/sce_scmouse.c	optional scemul kgc kgi
+dev/kgi/kgi_core.c		optional kgi
+dev/kgi/kgi_daemon.c		optional kgi
+dev/kgi/kgi_dpynull.c		optional kgi
+dev/kgi/kgi_syspci.c		optional kgi
+dev/kgi/kgi_sysatomic.c		optional kgi
+dev/kgi/kgi_sysmem.c		optional kgi
+dev/kgi/kgi_sysmutex.c		optional kgi
+dev/kgi/kgu_graphic.c		optional kgu kgi
+dev/kgi/kgu_resource.c		optional kgu kgi
+dev/kgi/kgu_command.c		optional kgu kgi
+dev/kgi/kgu_map.c		optional kgu kgi
+dev/kgi/kgu_accel.c		optional kgu kgi
+dev/kgi/kgu_mmio.c		optional kgu kgi
+dev/kii/kii_core.c		optional kgi
+dev/kii/kii_action.c		optional kgi
+dev/kii/kiu_event.c		optional kiu kgi
+dev/kip/kip_defaults.c		optional kip kgi
+dev/kip/kip_kbdriver.c		optional kip kgi
+dev/kip/kip_keymap.c		optional kip kgi
+dev/kip/kip_modifiers.c		optional kip kgi
+dev/kip/kip_symbols.c		optional kip kgi
 dev/le/am7990.c			optional le
 dev/le/am79900.c		optional le
 dev/le/if_le_pci.c		optional le pci
@@ -875,7 +915,7 @@
 dev/pci/pci_pci.c		optional pci
 dev/pci/pci_user.c		optional pci
 dev/pci/pcib_if.m		standard
-dev/pci/vga_pci.c		optional pci
+#dev/pci/vga_pci.c		optional pci
 dev/pdq/if_fea.c		optional fea eisa
 dev/pdq/if_fpa.c		optional fpa pci
 dev/pdq/pdq.c			optional nowerror fea eisa | fpa pci
@@ -2088,6 +2128,7 @@
 ufs/ufs/ufs_vnops.c		optional ffs
 vm/default_pager.c		standard
 vm/device_pager.c		standard
+vm/kgi_pager.c			standard
 vm/phys_pager.c			standard
 vm/redzone.c			optional DEBUG_REDZONE
 vm/swap_pager.c			standard
diff -Naur ./sys/conf/files.i386 /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/files.i386
--- ./sys/conf/files.i386	2008-02-23 21:08:16.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/files.i386	2008-02-23 19:59:46.000000000 +1100
@@ -221,7 +221,12 @@
 dev/ipmi/ipmi_smbios.c		optional ipmi
 dev/ipmi/ipmi_ssif.c		optional ipmi smbus
 dev/ipmi/ipmi_pci.c		optional ipmi pci
-dev/kbd/kbd.c			optional atkbd | sc | ukbd
+dev/kbd/kbd.c			optional atkbd | sc | ukbd | vt | kgi
+dev/kgy/kgy_disp.c		optional kgy kgi
+dev/kgy/kgy_mode.c		optional kgy kgi
+dev/kgy/kgy_text16.c		optional kgy kgi
+dev/kgy/kgy_splash.c		optional kgy kgi
+dev/kgim/kgim.c			optional kgim kgi
 dev/le/if_le_isa.c		optional le isa
 dev/mem/memutil.c		optional mem
 dev/mse/mse.c			optional mse
diff -Naur ./sys/conf/kmod.mk /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/kmod.mk
--- ./sys/conf/kmod.mk	2008-02-23 21:08:16.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/kmod.mk	2008-02-23 19:59:46.000000000 +1100
@@ -333,7 +333,8 @@
 	dev/sound/midi/synth_if.m dev/usb/usb_if.m isa/isa_if.m \
 	kern/bus_if.m kern/cpufreq_if.m kern/device_if.m kern/serdev_if.m \
 	libkern/iconv_converter_if.m opencrypto/cryptodev_if.m \
-	pc98/pc98/canbus_if.m
+	pc98/pc98/canbus_if.m pci/agp_if.m \
+	dev/kgc/scroller_if.m dev/kgc/render_if.m
 
 .for _srcsrc in ${MFILES}
 .for _ext in c h
diff -Naur ./sys/conf/options /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/options
--- ./sys/conf/options	2008-02-23 21:08:16.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/options	2008-02-23 19:59:46.000000000 +1100
@@ -513,6 +513,15 @@
 NETATALKDEBUG		opt_atalk.h
 SI_DEBUG		opt_debug_si.h
 
+# KGI options
+KGI_COMPAT		opt_kgi.h
+KGI_DBG_LEVEL		opt_kgi.h
+KGC_DEFAULT_FONT	opt_kgi.h
+KGC_RENDER_BACKGROUND	opt_kgi.h
+KGC_RENDER_1024x768	opt_kgi.h
+KGC_RENDER_800x600	opt_kgi.h
+KGC_RENDER_16BITS	opt_kgi.h
+
 # Fb options
 FB_DEBUG		opt_fb.h
 FB_INSTALL_CDEV		opt_fb.h
diff -Naur ./sys/conf/options.i386 /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/options.i386
--- ./sys/conf/options.i386	2008-02-23 21:08:16.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/options.i386	2008-02-23 19:59:46.000000000 +1100
@@ -91,6 +91,10 @@
 
 VESA
 VESA_DEBUG		opt_vesa.h
+  
+KGI_NOSPLASH		opt_kgi.h
+CONFIG_FB_ATY_GX	opt_kgi.h
+CONFIG_FB_ATY_CT	opt_kgi.h
 
 PSM_DEBUG		opt_psm.h
 PSM_HOOKRESUME		opt_psm.h
diff -Naur ./sys/conf/updep.sh /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/updep.sh
--- ./sys/conf/updep.sh	1970-01-01 10:00:00.000000000 +1000
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/updep.sh	2008-02-23 19:59:46.000000000 +1100
@@ -0,0 +1,12 @@
+#! /bin/sh
+#
+# updep <ofile> <cfile>
+#
+# $FreeBSD$
+#
+
+OFILE=$1; shift;
+CFILE=$1; shift;
+
+sed -e "s#^.*: ${CFILE}#${OFILE}: ${CFILE}#" .newdep > .depend
+mv .depend .newdep
diff -Naur ./sys/dev/fb/fb.c /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/dev/fb/fb.c
--- ./sys/dev/fb/fb.c	2008-02-23 21:08:23.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/dev/fb/fb.c	2008-02-23 19:59:54.000000000 +1100
@@ -533,6 +533,7 @@
 	{ KD_VGA,	"VGA" },
 	{ KD_PC98,	"PC-98x1" },
 	{ KD_TGA,	"TGA" },
+	{ KD_KGI,	"KGI" },
 	{ -1,		"Unknown" },
     };
     int i;
@@ -608,6 +609,7 @@
 		{ FBTYPE_VGA,		KD_VGA },
 		{ FBTYPE_PC98,		KD_PC98 },
 		{ FBTYPE_TGA,		KD_TGA },
+		{ FBTYPE_KGI,		KD_KGI },
 	};
 	int i;
diff -Naur ./sys/i386/isa/vesa.c /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/i386/isa/vesa.c
--- ./sys/i386/isa/vesa.c	2008-02-23 21:08:28.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/i386/isa/vesa.c	2008-02-23 20:00:02.000000000 +1100
@@ -851,7 +851,9 @@
 
 /* entry points */
 
-static int
+extern int vesa_configure(int flags);
+
+int
 vesa_configure(int flags)
 {
 	video_adapter_t *adp;
diff -Naur ./sys/isa/vga_isa.c /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/isa/vga_isa.c
--- ./sys/isa/vga_isa.c	2008-02-23 21:08:29.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/isa/vga_isa.c	2008-02-23 20:00:03.000000000 +1100
@@ -30,6 +30,7 @@
 #include "opt_vga.h"
 #include "opt_fb.h"
 #include "opt_syscons.h"	/* should be removed in the future, XXX */
+#include "opt_kgi.h"
 
 #include <sys/param.h>
 #include <sys/systm.h>
@@ -106,10 +107,12 @@
 	device_set_desc(dev, "Generic ISA VGA");
 	error = vga_probe_unit(device_get_unit(dev), &adp, device_get_flags(dev));
 	if (error == 0) {
+#ifndef KGI_COMPAT
 		bus_set_resource(dev, SYS_RES_IOPORT, 0,
 				 adp.va_io_base, adp.va_io_size);
 		bus_set_resource(dev, SYS_RES_MEMORY, 0,
 				 adp.va_mem_base, adp.va_mem_size);
+#endif
 #if 0
 		isa_set_port(dev, adp.va_io_base);
 		isa_set_portsize(dev, adp.va_io_size);
@@ -125,8 +128,9 @@
 {
 	vga_softc_t *sc;
 	int unit;
-	int rid;
 	int error;
+#ifndef KGI_COMPAT
+	int rid;
 
 	unit = device_get_unit(dev);
 	sc = device_get_softc(dev);
@@ -137,6 +141,10 @@
 	rid = 0;
 	bus_alloc_resource(dev, SYS_RES_MEMORY, &rid,
 				 0, ~0, 0, RF_ACTIVE | RF_SHAREABLE);
+#endif
+
+	unit = device_get_unit(dev);
+	sc = device_get_softc(dev);
 
 	error = vga_attach_unit(unit, sc, device_get_flags(dev));
 	if (error)
diff -Naur ./sys/sys/conf.h /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/sys/conf.h
--- ./sys/sys/conf.h	2008-02-23 21:08:36.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/sys/conf.h	2008-02-23 20:00:09.000000000 +1100
@@ -171,6 +171,7 @@
 #define D_MMAP_ANON	0x00100000	/* special treatment in vm_mmap.c */
 #define D_PSEUDO	0x00200000	/* make_dev() can return NULL */
 #define D_NEEDGIANT	0x00400000	/* driver want Giant */
+#define D_KGI_PAGING	0x00800000	/* is a KGI pager backend */
 
 /*
  * Version numbers.
diff -Naur ./sys/sys/fbio.h /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/sys/fbio.h
--- ./sys/sys/fbio.h	2008-02-23 21:08:36.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/sys/fbio.h	2008-02-23 20:00:09.000000000 +1100
@@ -85,8 +85,9 @@
 #define	FBTYPE_MDICOLOR		28	/* cg14 */
 #define	FBTYPE_TCXCOLOR		29	/* SUNW,tcx */
 #define	FBTYPE_CREATOR		30
+#define	FBTYPE_KGI		31
 
-#define	FBTYPE_LASTPLUSONE	31	/* max number of fbs (change as add) */
+#define	FBTYPE_LASTPLUSONE	32	/* max number of fbs (change as add) */
 
 /*
  * Frame buffer descriptor as returned by FBIOGTYPE.
@@ -269,6 +270,7 @@
 #define V_INFO_GRAPHICS	(1 << 1)
 #define V_INFO_LINEAR	(1 << 2)
 #define V_INFO_VESA	(1 << 3)
+#define V_INFO_KGI	(1 << 4)
     int			vi_width;
     int			vi_height;
     int			vi_cwidth;
@@ -314,6 +316,7 @@
 #define KD_PC98		6		/* PC-98 display */
 #define KD_TGA		7		/* TGA */
 #define KD_TGA2		8		/* TGA2 */
+#define KD_KGI		9		/* KGI */
     char		*va_name;
     int			va_unit;
     int			va_minor;
@@ -331,6 +334,7 @@
 #define V_ADP_INITIALIZED (1 << 17)
 #define V_ADP_REGISTERED (1 << 18)
 #define V_ADP_ATTACHED	(1 << 19)
+#define V_ADP_KGI	(1 << 20)
     vm_offset_t		va_io_base;
     int			va_io_size;
     vm_offset_t		va_crtc_addr;
diff -Naur ./sys/sys/kernel.h /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/sys/kernel.h
--- ./sys/sys/kernel.h	2008-02-23 21:08:36.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/sys/kernel.h	2008-02-23 20:00:09.000000000 +1100
@@ -111,6 +111,7 @@
 	SI_SUB_MTX_POOL_DYNAMIC	= 0x1AC0000,	/* dynamic mutex pool */
 	SI_SUB_LOCK		= 0x1B00000,	/* various locks */
 	SI_SUB_EVENTHANDLER	= 0x1C00000,	/* eventhandler init */
+ 	SI_SUB_KGI		= 0x1D00000,	/* KGI */
 	SI_SUB_KLD		= 0x2000000,	/* KLD and module setup */
 	SI_SUB_CPU		= 0x2100000,	/* CPU resource(s)*/
 	SI_SUB_MAC		= 0x2180000,	/* TrustedBSD MAC subsystem */
@@ -162,6 +163,7 @@
 	SI_SUB_KTHREAD_VM	= 0xe800000,	/* vm daemon*/
 	SI_SUB_KTHREAD_BUF	= 0xea00000,	/* buffer daemon*/
 	SI_SUB_KTHREAD_UPDATE	= 0xec00000,	/* update daemon*/
+	SI_SUB_KTHREAD_KGI	= 0xed00000,	/* kgi daemon*/
 	SI_SUB_KTHREAD_IDLE	= 0xee00000,	/* idle procs*/
 	SI_SUB_SMP		= 0xf000000,	/* start the APs*/
 	SI_SUB_RUN_SCHEDULER	= 0xfffffff	/* scheduler*/
diff -Naur ./sys/sys/mman.h /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/sys/mman.h
--- ./sys/sys/mman.h	2008-02-23 21:08:36.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/sys/mman.h	2008-02-23 20:00:09.000000000 +1100
@@ -59,6 +59,7 @@
  */
 #define	MAP_SHARED	0x0001		/* share changes */
 #define	MAP_PRIVATE	0x0002		/* changes are private */
+#define	MAP_KGI		0x0004		/* KGI mmapping */
 #if __BSD_VISIBLE
 #define	MAP_COPY	MAP_PRIVATE	/* Obsolete */
 #endif
diff -Naur ./sys/vm/vm_fault.c /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_fault.c
--- ./sys/vm/vm_fault.c	2008-02-23 21:08:32.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_fault.c	2008-02-23 20:00:06.000000000 +1100
@@ -484,7 +484,8 @@
 			      fs.pindex < fs.entry->lastr + VM_FAULT_READ)) &&
 			    (fs.first_object == fs.object ||
 			     (is_first_object_locked = VM_OBJECT_TRYLOCK(fs.first_object))) &&
-			    fs.first_object->type != OBJT_DEVICE) {
+			    (fs.first_object->type != OBJT_DEVICE) &&
+			    (fs.first_object->type != OBJT_KGI)) {
 				vm_pindex_t firstpindex, tmppindex;
 
 				if (fs.first_pindex < 2 * VM_FAULT_READ)
diff -Naur ./sys/vm/vm.h /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm.h
--- ./sys/vm/vm.h	2008-02-23 21:08:32.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm.h	2008-02-23 20:00:06.000000000 +1100
@@ -81,7 +81,7 @@
 #define	VM_PROT_DEFAULT		VM_PROT_ALL
 
 enum obj_type { OBJT_DEFAULT, OBJT_SWAP, OBJT_VNODE, OBJT_DEVICE, OBJT_PHYS,
-		OBJT_DEAD };
+		OBJT_KGI, OBJT_DEAD };
 typedef u_char objtype_t;
 
 union vm_map_object;
diff -Naur ./sys/vm/vm_mmap.c /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_mmap.c
--- ./sys/vm/vm_mmap.c	2008-02-23 21:08:32.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_mmap.c	2008-02-23 20:00:06.000000000 +1100
@@ -212,7 +212,7 @@
 	struct pmckern_map_in pkm;
 #endif
 	struct file *fp;
-	struct vnode *vp;
+	struct vnode *vp = NULL;
 	vm_offset_t addr;
 	vm_size_t size, pageoff;
 	vm_prot_t prot, maxprot;
@@ -357,6 +357,13 @@
 		} else if (vp->v_type != VCHR || (fp->f_flag & FWRITE) != 0) {
 			maxprot |= VM_PROT_WRITE;
 		}
+  
+		/* Recognize a KGI pager backend */
+		if ((vp->v_type == VCHR) && 
+		    (vp->v_rdev->si_devsw->d_flags & D_KGI_PAGING)) {
+			flags |= MAP_KGI;
+		}
+
 		handle = (void *)vp;
 		handle_type = OBJT_VNODE;
 	}
@@ -1169,7 +1176,10 @@
 		type = OBJT_VNODE;
 		handle = vp;
 	} else if (vp->v_type == VCHR) {
-		type = OBJT_DEVICE;
+		if (flags & MAP_KGI)
+			type = OBJT_KGI;
+		else
+			type = OBJT_DEVICE;
 		handle = vp->v_rdev;
 
 		/* XXX: lack thredref on device */
@@ -1228,7 +1238,7 @@
 	}
 	obj = vm_pager_allocate(type, handle, objsize, prot, foff);
 	if (obj == NULL) {
-		error = (type == OBJT_DEVICE ? EINVAL : ENOMEM);
+		error = ((type == OBJT_DEVICE || type ==OBJT_KGI) ? EINVAL : ENOMEM);
 		goto done;
 	}
 	*objp = obj;
diff -Naur ./sys/vm/vm_object.h /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_object.h
--- ./sys/vm/vm_object.h	2008-02-23 21:08:32.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_object.h	2008-02-23 20:00:06.000000000 +1100
@@ -132,6 +132,15 @@
 		struct {
 			int swp_bcount;
 		} swp;
+
+		/*
+		 * KGI pager
+		 *
+		 *	kgip_pglist - list of allocated pages
+		 */
+		struct {
+			TAILQ_HEAD(, vm_page) kgip_pglist;
+		} kgip;
 	} un_pager;
 };
diff -Naur ./sys/vm/vm_page.h /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_page.h
--- ./sys/vm/vm_page.h	2008-02-23 21:08:32.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_page.h	2008-02-23 20:00:06.000000000 +1100
@@ -103,6 +103,7 @@
 struct vm_page {
 	TAILQ_ENTRY(vm_page) pageq;	/* queue info for FIFO queue or free list (P) */
 	TAILQ_ENTRY(vm_page) listq;	/* pages in same object (O) 	*/
+	TAILQ_ENTRY(vm_page) kgiq;	/* pages in same kgi buf (O) 	*/
 	struct vm_page *left;		/* splay tree link (O)		*/
 	struct vm_page *right;		/* splay tree link (O)		*/
diff -Naur ./sys/vm/vm_pager.c /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_pager.c
--- ./sys/vm/vm_pager.c	2008-02-23 21:08:32.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_pager.c	2008-02-23 20:00:06.000000000 +1100
@@ -163,6 +163,7 @@
 	&vnodepagerops,		/* OBJT_VNODE */
 	&devicepagerops,	/* OBJT_DEVICE */
 	&physpagerops,		/* OBJT_PHYS */
+	&kgipagerops,		/* OBJT_KGI */
 	&deadpagerops		/* OBJT_DEAD */
 };
diff -Naur ./sys/vm/vm_pager.h /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_pager.h
--- ./sys/vm/vm_pager.h	2008-02-23 21:08:32.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/vm/vm_pager.h	2008-02-23 20:00:06.000000000 +1100
@@ -71,6 +71,7 @@
 extern struct pagerops vnodepagerops;
 extern struct pagerops devicepagerops;
 extern struct pagerops physpagerops;
+extern struct pagerops kgipagerops;
 
 /*
  * get/put return values
diff -Naur ./usr.sbin/moused/moused.c /usr/home/agh/Dev/KGI/kgi8/kgi8-src/usr.sbin/moused/moused.c
--- ./usr.sbin/moused/moused.c	2008-02-23 21:08:07.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/usr.sbin/moused/moused.c	2008-02-23 20:00:35.000000000 +1100
@@ -1003,8 +1003,10 @@
     int c;
     int i;
 
-    if ((rodent.cfd = open("/dev/consolectl", O_RDWR, 0)) == -1)
-	logerr(1, "cannot open /dev/consolectl");
+	if ((rodent.cfd = open("/dev/consolectl", O_RDWR, 0)) == -1) {
+		if ((rodent.cfd = open("/dev/consolectl1", O_RDWR, 0)) == -1)
+			logerrx(1, "cannot open any /dev/consolectl");
+	}
 
     if (!nodaemon && !background) {
 	pfh = pidfile_open(pidfile, 0600, &mpid);
diff -Naur ./sys/conf/kern.mk /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/kern.mk
--- ./sys/conf/kern.mk	2008-02-23 21:08:16.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/conf/kern.mk	2008-02-23 19:59:46.000000000 +1100
@@ -12,7 +12,7 @@
 .else
 CWARNFLAGS?=	-Wall -Wredundant-decls -Wnested-externs -Wstrict-prototypes \
 		-Wmissing-prototypes -Wpointer-arith -Winline -Wcast-qual \
-		${_wundef} ${_Wno_pointer_sign} -fformat-extensions
+		${_wundef} ${_Wno_pointer_sign} -fformat-extensions -fgnu89-inline
 .if !defined(WITH_GCC3)
 _Wno_pointer_sign=-Wno-pointer-sign
 .endif
diff -Naur ./sys/dev/kbd/kbd.c /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/dev/kbd/kbd.c
--- ./sys/dev/kbd/kbd.c	2008-02-23 21:08:24.000000000 +1100
+++ /usr/home/agh/Dev/KGI/kgi8/kgi8-src/sys/dev/kbd/kbd.c	2008-02-23 19:59:56.000000000 +1100
@@ -180,6 +180,8 @@
 	return (0);
 }
 
+void kip_kbd_register(keyboard_t *kbd, int index);
+
 /* register a keyboard and associate it with a function table */
 int
 kbd_register(keyboard_t *kbd)
@@ -222,7 +224,7 @@
 				kbdd_ioctl(mux, KBADDKBD, (caddr_t) &ki);
 			}
 
-			return (index);
+			goto found;
 		}
 	}
 	SET_FOREACH(list, kbddriver_set) {
@@ -239,11 +241,15 @@
 				kbdd_ioctl(mux, KBADDKBD, (caddr_t) &ki);
 			}
 
-			return (index);
+			goto found;
 		}
 	}
 
 	return (-1);
+
+found:
+	kip_kbd_register(kbd, index);                         /* XXX hardwired */
+	return (index);
 }
 
 int